### PR TITLE
Type Equality Checking

### PR DESCRIPTION
**Pull request by ZacharyPatten**
long story short... when dealing with generics... you can speed up equality checking via caching

### CODE REVIEW

File: Benchmarks/TypeEqualityChecking/TypeEqualityCheckingBenchmarks.cs [hunk 1]
```diff
@@ -0,0 +1,130 @@
...
+			temp = TypeOfEqualsGeneric<string, int>();
+			temp = TypeOfEqualsGeneric<int, string>();
+			temp = TypeOfEqualsGeneric<double, int>();
+			temp = TypeOfEqualsGeneric<decimal, int>();
+			temp = TypeOfEqualsGeneric<byte, int>();
+			temp = TypeOfEqualsGeneric<short, int>();
+		}
+
+		[Benchmark]
+		public void TypeOfEqualsGenericInlined()
```
Comment (WhiteBlackGoose) [inline Benchmarks/TypeEqualityChecking/TypeEqualityCheckingBenchmarks.cs:62]: So I added this as as an inlined version of the generic method

File: Benchmarks/TypeEqualityChecking/TypeEqualityCheckingBenchmarks.cs [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+			temp = typeof(double) == typeof(int);
+			temp = typeof(decimal) == typeof(int);
+			temp = typeof(byte) == typeof(int);
+			temp = typeof(short) == typeof(int);
+		}
+
+		static bool TypeOfEqualsGeneric<A, B>() => typeof(A) == typeof(B);
+
+		[MethodImpl(MethodImplOptions.AggressiveInlining)]
+		static bool TypeOfEqualsGenericInlined<A, B>() => typeof(A) == typeof(B);
```
Comment (WhiteBlackGoose) [inline Benchmarks/TypeEqualityChecking/TypeEqualityCheckingBenchmarks.cs]: We force inlining as JIT misses this optimization for some reason

File: Benchmarks/TypeEqualityChecking/README.md [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+  DefaultJob : .NET Core 5.0.0 (CoreCLR 5.0.20.47505, CoreFX 5.0.20.47505), X64 RyuJIT
+
+
+```
+
+|                     Method |      Mean |     Error |    StdDev |
+|--------------------------- |----------:|----------:|----------:|
+|               TypeOfEquals |  2.498 ns | 0.0904 ns | 0.1630 ns |
+|        TypeOfEqualsGeneric | 22.482 ns | 0.4369 ns | 0.4086 ns |
+| TypeOfEqualsGenericInlined |  2.537 ns | 0.0883 ns | 0.0826 ns |
```
Comment (WhiteBlackGoose) [inline Benchmarks/TypeEqualityChecking/README.md]: That works as fast as `TypeOfEquals`, and it should work so

File: Benchmarks/TypeEqualityChecking/TypeEqualityCheckingBenchmarks.cs [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+			temp = TypeOfEqualsGenericInlined<string, int>();
+			temp = TypeOfEqualsGenericInlined<int, string>();
+			temp = TypeOfEqualsGenericInlined<double, int>();
+			temp = TypeOfEqualsGenericInlined<decimal, int>();
+			temp = TypeOfEqualsGenericInlined<byte, int>();
+			temp = TypeOfEqualsGenericInlined<short, int>();
+		}
+
+		struct Operator_is_struct<A, B> { }
+		static bool Operator_is<A, B>() => default(Operator_is_struct<A, B>) is Operator_is_struct<A, A>;
```
Comment (WhiteBlackGoose) [inline Benchmarks/TypeEqualityChecking/TypeEqualityCheckingBenchmarks.cs]: I don't think we need it anymore, to be honest. I'd remove it

**Review by WhiteBlackGoose — COMMENTED** [review]
So, seems like you can rename this benchmark from type checking to testing the JIT's mechanism for inlining generic methods :)

File: Benchmarks/TypeEqualityChecking/TypeEqualityCheckingBenchmarks.cs (hunk 4 above)
Comment (ZacharyPatten) [reply Benchmarks/TypeEqualityChecking/TypeEqualityCheckingBenchmarks.cs]: It depends on what the goal of the benchmark is. Personally I think it should stay on the topic of type equality checking (in which case keeping the benchmark is good imo). If you think we should make the benchmark represent something else (such as examples where the JIT can't inline) then that would IMO require a lot more examples.

**Comment by ZacharyPatten** [comment]
I think this PR is ready to merge now.

**Pull request merged**